### DISCUSSION TITLE
missing get_recent_deployments file

### DISCUSSION THREAD

**Discussion by PtSik** [Cyfrin/foundry-full-course-cu#1801] [General]
After installing foundry-devops I dont have get_recent_deployments file in the devops folder. I fixed it somehow by creating couple functions but would like to avoid it in the future.

**Comment by heheboii11** [comment]
Post the exact error you are facing

**Comment by EngrPips** [comment]
Hello @PtSik, providing more details might be helpful to help you.

**Comment by ghost** [comment]
Hello @EngrPips I just installed the foundry DevOps and it's missing the script though I haven't continued I can see there would be an issue moving ahead and also some lessons further down seem to use the same method to get the latest deployment any suggestions please.
@PtSik please what was your solution to this I thought the foundry cheatsheet would have something for this but I haven't been able to find one.

**Reply by EngrPips** [reply]
Hello @kdshola, Below is Foundry native way of doing it. I have seen it work for people. Let me know if it works for you.

```js
import {stdJson} from "forge-std/StdJson.sol";
import {Strings} from "@openzeppelin/contracts/utils/Strings.sol";

function getDeployedContractAddress() private view returns (address) {
        string memory path = string.concat(
            vm.projectRoot(),
            "/broadcast/DeployYourContract.s.sol/",
            Strings.toString(block.chainid),
            "/run-latest.json"
        );
        string memory json = vm.readFile(path);
        bytes memory contractAddress = stdJson.parseRaw(
            json,
            ".transactions[0].contractAddress"
        );
        return (bytesToAddress(contractAddress));
    }

    function bytesToAddress(
        bytes memory bys
    ) private pure returns (address addr) {
        assembly {
            addr := mload(add(bys, 32))
        }
    }
```

**Reply by ghost** [reply]
Thanks I'll let you know is there any link to materials that can help me understand what the code does.

**Reply by EngrPips** [reply]
Honestly, there isn't any link. Someone like you, taking this course, runs into a similar issue, and they went ahead and set up this function and shared it with me in the discussion that they opened to get help. I mean, I just patiently went through the lines one by one, and I was able to understand what the code needed to do. Maybe you can do the same, too.

**Reply by ghost** [reply]
I can understand that it's reading the broadcast directory and the latest.json file  which us always overwritten with the latest deployed contract's information I can sort of understand how it would work I guess I am just not familiar with those functions and the mload opcode.

**Reply by EngrPips** [reply]
Oh, I see. Try using it, and let's see if it works for you. Then, we can dive deep into the low-level line of code.

**Reply by EngrPips** [reply]
But the latest version of Cyfrin foundry-devops does not have a file name `get-recent-deployment`; instead, it is a function inside the `DevOps` contract. See the image below.

<img width="1440" alt="Screenshot 2024-06-09 at 16 50 34" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/95760918/ce9ed45b-7b2c-46da-b8b5-2a93bb9ba786">

**Reply by ghost** [reply]
I'll try this out and if it doesn't work I'll use the method you provided above but that may be tomorrow.

**Reply by EngrPips** [reply]
Alright, Friend. I am looking forward to hearing from you.

**Reply by EngrPips** [reply]
Hey @kdshola, I used Cyfrin `foundry-devops` just now, and it still works fine.

<img width="1440" alt="Screenshot 2024-06-09 at 17 55 33" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/95760918/98ec0668-832a-4302-aa6a-188b1a506c2d">

**Reply by ghost** [reply]
Ok thanks @EngrPips I guess they updated the contract because of the possible security issues with executing bash commands does the foundry.toml file require any change or just the setups in the foundry-devops repo just it.

**Reply by EngrPips** [reply]
Yeah, you need only the below in `foundry.toml`

`fs_permissions = [{ access = "read", path = "./broadcast" }]`

**Reply by ghost** [reply]
Thanks I've tried it and it works @EngrPips

**Reply by EngrPips** [reply]
Awesome, Keep crushing it.
